### PR TITLE
[receiver/hostmetrics] Divide by logical cores when calculating process.cpu.utilization

### LINKED ISSUE #31368: [receiver/hostmetrics] The value of process.cpu.utilization may exceed 1
### Component(s)

receiver/hostmetrics

## Description

The `process.cpu.utilization` metric is expected to be a value between 0 and 1, but it can be greater than 1.

My guess here is that process.cpu.utilization does not properly divide by the number of system cores, so the value can actually be between 0 - ${num_cores}

## Steps to Reproduce

1. Run a process with heavy load on a multi-core machine
2. Scrape with the hostmetric receiver
3. Note that process metrics may exceed 1

## Expected Result
No metrics exceed 1 (in fact, I'd expect the sum of all processes to be < 1)

## Actual Result

I'm getting a metric of 9.5 (this process was taking ~950% cpu in my activity monitor)
```json
{
  "resource": {
    "attributes": [
      { "key": "process.pid", "value": { "intValue": "89633" } },
      { "key": "process.parent_pid", "value": { "intValue": "89617" } },
      {
        "key": "process.executable.name",
        "value": { "stringValue": "load" }
      },
      {
        "key": "process.executable.path",
        "value": {
          "stringValue": "/var/folders/qn/13392c0n4rl4j4cplc65_66h0000gn/T/go-build2476508403/b001/exe/load"
        }
      },
      {
        "key": "process.command",
        "value": {
          "stringValue": "/var/folders/qn/13392c0n4rl4j4cplc65_66h0000gn/T/go-build2476508403/b001/exe/load"
        }
      },
      {
        "key": "process.command_line",
        "value": {
          "stringValue": "/var/folders/qn/13392c0n4rl4j4cplc65_66h0000gn/T/go-build2476508403/b001/exe/load"
        }
      },
      {
        "key": "process.owner",
        "value": { "stringValue": "brandonjohnson" }
      },
      { "key": "host.name", "value": { "stringValue": "Brandons-MBP" } },
      { "key": "os.type", "value": { "stringValue": "darwin" } }
    ]
  },
  "scopeMetrics": [
    {
      "scope": {
        "name": "otelcol/hostmetricsreceiver/process",
        "version": "v1.44.0"
      },
      "metrics": [
        {
          "name": "process.cpu.utilization",
          "description": "Percentage of total CPU time used by the process since last scrape, expressed as a value between 0 and 1. On the first scrape, no data point is emitted for this metric.",
          "unit": "1",
          "gauge": {
            "dataPoints": [
              {
                "attributes": [
                  { "key": "state", "value": { "stringValue": "user" } }
                ],
                "startTimeUnixNano": "1708529361162000000",
                "timeUnixNano": "1708529423825350000",
                "asDouble": 9.532629437655167
              },
              {
                "attributes": [
                  { "key": "state", "value": { "stringValue": "system" } }
                ],
                "startTimeUnixNano": "1708529361162000000",
                "timeUnixNano": "1708529423825350000",
                "asDouble": 0.07040537266051189
              },
              {
                "attributes": [
                  { "key": "state", "value": { "stringValue": "wait" } }
                ],
                "startTimeUnixNano": "1708529361162000000",
                "timeUnixNano": "1708529423825350000",
                "asDouble": 0
              }
            ]
          }
        }
      ]
    }
  ],
  "schemaUrl": "https://opentelemetry.io/schemas/1.9.0"
}
```

### Collector version

v0.94.0

## Environment
OS: macOS 14.2.1
Compiler(if manually compiled): go 1.22

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

I used this quick go program to quickly generate load:
```go
package main

import "time"

func main() {
	for i := 0; i < 100; i++ {
		go func() {
			v := 0
			for {
				v++
			}
		}()
	}

	time.Sleep(3 * time.Minute)
}

```

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:**
When calculating the process.cpu.utilization metric, values over 1 were possible since the number of cores was not taken into account (a single process may run on multiple logical cores, this effectively multplying the maximum amount of CPU time the process may take).

This PR adds a division by the number of logical cores to the calculation for cpu utilization.

**Link to tracking Issue:** Closes #31368

**Testing:**
* Added some unit tests
* Tested locally on my system with the program I posted in the issue:

```json
{
  "name": "process.cpu.utilization",
  "description": "Percentage of total CPU time used by the process since last scrape, expressed as a value between 0 and 1. On the first scrape, no data point is emitted for this metric.",
  "unit": "1",
  "gauge": {
    "dataPoints": [
      {
        "attributes": [{ "key": "state", "value": { "stringValue": "user" } }],
        "startTimeUnixNano": "1708562810521000000",
        "timeUnixNano": "1708562890771386000",
        "asDouble": 0.8811268516953904
      },
      {
        "attributes": [
          { "key": "state", "value": { "stringValue": "system" } }
        ],
        "startTimeUnixNano": "1708562810521000000",
        "timeUnixNano": "1708562890771386000",
        "asDouble": 0.0029471002907659667
      },
      {
        "attributes": [{ "key": "state", "value": { "stringValue": "wait" } }],
        "startTimeUnixNano": "1708562810521000000",
        "timeUnixNano": "1708562890771386000",
        "asDouble": 0
      }
    ]
  }
}
```

In activity monitor, this process was clocking in around ~1000% - ~1100% cpu, on my machine that has 12 logical cores. So the value of around 90% total utilization seems correct here.

**Documentation:**
N/A

### CODE REVIEW

**Review by djaglowski — APPROVED** [review]
This PR would be correct according to the [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/096596b200faeefdca7e27e2ef60113f908eafbe/docs/system/process-metrics.md#metric-processcpuutilization) which define the metric as: 

"Difference in process.cpu.time since the last measurement, divided by the elapsed time *and number of CPUs available to the process*."

**Review by braydonk — APPROVED** [review]
I would definitely wait for @dmitryax to chime in, as part of system semantic conventions work he will be going through and clearing up all the definitions of CPU utilization and process.

Personally I think it's fine to merge this PR because based on the description of this metric in the current metadata.yaml, the existing implementation is categorically wrong and this fixes it as far as I can tell. I'm a bit fuzzy on whether this technically constitutes a breaking change, since we are now fixing something that was incorrect but technically users could be relying on that incorrect number. I'm not sure on that judgement call.

On a personal level I actually am not a huge fan of the described behaviour of this metric. At least in Linux-land, a process' CPU utilization is often interpreted as the percentage of a single core, thus the percentage actually could be over 1 if the process uses more than 1 full core. I discussed this in the last system semconv meeting, and I will open an issue about it for us to work out there.

All that aside, my opinion is that this PR should be merged to match the description, and the semconv working group should make a decision what we want this metric to be in the final process semantic conventions. Definitely wait for Dmitrii to weigh in though.

**Review by andrzej-stencel — APPROVED** [review]
I agree that this PR aligns the receiver's behavior with the current description in the [Semantic Conventions for OS Process Metrics](https://github.com/open-telemetry/semantic-conventions/blob/v1.24.0/docs/system/process-metrics.md).

The only caveat is that the OS process convention itself is apparently in "Experimental" status. We probably don't want to change this behavior twice. ;)

Perhaps worth marking this as a `breaking` change to make it more visible to users who might be relying on the current behavior.

**Comment by BinaryFissionGames** [comment]
Bump @dmitryax - Would like to get your thoughts on this one!

**Review by dmitryax — APPROVED** [review]
I think it's the right fix given that we don't have an attribute for a particular core as we have for `system.cpu.utilization `

**Comment by dmitryax** [comment]
But, given that it's a widely used component, should we make the change with a feature gate? @djaglowski @astencel-sumo @braydonk

**Comment by astencel-sumo** [comment]
> But, given that it's a widely used component, should we make the change with a feature gate? @djaglowski @astencel-sumo @braydonk

Yeah I'm afraid I agree. :wink: It's always additional work adding the feature gate and changing it in later stages, but probably it's justified in this case. This might be quite a disruptive change for users relying on this metric.

**Comment by BinaryFissionGames** [comment]
@dmitryax I've updated the PR to add a new featuregate, and optionally normalize the metric based on the feature gate.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/ucal/cpu_utilization_calculator.go [hunk 1]
```diff
@@ -7,9 +7,18 @@ import (
 	"time"
 
 	"github.com/shirou/gopsutil/v3/cpu"
+	"go.opentelemetry.io/collector/featuregate"
 	"go.opentelemetry.io/collector/pdata/pcommon"
 )
 
+var normalizeCPURatioFeatureGate = featuregate.GlobalRegistry().MustRegister(
+	"receiver.hostmetricsreceiver.normalizeProcessCpuUtilization",
```
Comment (astencel-sumo) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/ucal/cpu_utilization_calculator.go:15]:
```suggestion
	"receiver.hostmetrics.normalizeProcessCpuUtilization",
```

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/ucal/cpu_utilization_calculator.go [hunk 2]
```diff
@@ -7,9 +7,18 @@ import (
...
+	featuregate.StageAlpha,
+	featuregate.WithRegisterDescription("When enabled, normalizes the process.cpu.utilization metric onto the interval [0-1] by dividing by the metric number of logical processors."),
```
Comment (astencel-sumo) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/ucal/cpu_utilization_calculator.go:17]: Did you mean this?
```suggestion
	featuregate.WithRegisterDescription("When enabled, normalizes the process.cpu.utilization metric onto the interval [0-1] by dividing the value by the number of logical processors."),
```

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/ucal/cpu_utilization_calculator.go [hunk 3]
```diff
@@ -7,9 +7,18 @@ import (
 	"time"
 
 	"github.com/shirou/gopsutil/v3/cpu"
+	"go.opentelemetry.io/collector/featuregate"
 	"go.opentelemetry.io/collector/pdata/pcommon"
 )
 
+var normalizeCPURatioFeatureGate = featuregate.GlobalRegistry().MustRegister(
```
Comment (astencel-sumo) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/ucal/cpu_utilization_calculator.go:14]:
```suggestion
var normalizeProcessCpuUtilizationFeatureGate = featuregate.GlobalRegistry().MustRegister(
```

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/ucal/cpu_utilization_calculator_test.go [hunk 4]
```diff
@@ -108,9 +179,26 @@ func Test_cpuUtilization(t *testing.T) {
...
 
-	actualUtilization := cpuUtilization(startStat, startTime, endStat, halfSecondLater)
+	actualUtilization := cpuUtilization(1, startStat, startTime, endStat, halfSecondLater)
 	assert.InDelta(t, expectedUtilization.User, actualUtilization.User, 0.00001)
 	assert.InDelta(t, expectedUtilization.System, actualUtilization.System, 0.00001)
 	assert.InDelta(t, expectedUtilization.Iowait, actualUtilization.Iowait, 0.00001)
 
 }
+
+func setNormalizeCPURatioFeatureGate(t *testing.T, val bool) {
```
Comment (astencel-sumo) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/ucal/cpu_utilization_calculator_test.go:189]:
```suggestion
func setNormalizeProcessCpuUtilizationFeatureGate(t *testing.T, val bool) {
```

**Comment by astencel-sumo** [comment]
@BinaryFissionGames , thanks for adding the feature gate. I've added a couple comments to make naming more consistent.

Can you also please add documentation describing the behavior and the schedule of the feature gate. See e.g. here for inspiration: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.87.0/extension/storage/filestorage/README.md#extensionfilestoragereplaceunsafecharacters.

File: receiver/hostmetricsreceiver/README.md [hunk 5]
```diff
@@ -192,3 +192,18 @@ Currently, the hostmetrics receiver does not set any Resource attributes on the
...
+
+When enabled, normalizes the `process.cpu.utilization` metric onto the interval [0-1] by dividing the value by the number of logical processors. With this feature gate disabled, the value of the `process.cpu.utilization` metric may exceed 1.
+
+For example, if you have 4 logical cores on your system, and a process is occupying 2 logical cores for an entire scrape interval, with this feature gate disabled a `process.cpu.utilization` metric will be emitted with a value of 2.  if this feature gate is enabled in the same scenario, the value of the emitted metric will be 0.5.
+
+The schedule for this feature gate is:
+- Introduced in v0.97.0 (March 2024) as `alpha` - disabled by default.
+- Moved to `beta` in v0.99.0 (April 2024) - enabled by default.
+- Moved to `stable` in v0.101.0 (May 2024) - cannot be disabled.
+- Removed three releases after `stable`.
```
Comment (BinaryFissionGames) [inline receiver/hostmetricsreceiver/README.md:209]: Let me know if this schedule seems too compressed. Tried to go with the 2 release per advancement approach.

**Review by andrzej-stencel — APPROVED** [review]
Love it! Thank you @BinaryFissionGames :clap:

**Review by djaglowski — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

File: receiver/hostmetricsreceiver/README.md [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
+| [Code Owners](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#becoming-a-code-owner) | [@dmitryax](https://www.github.com/dmitryax), [@braydonk](https://www.github.com/braydonk)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/README.md:9]: Please revert this change to make CI pass. This is auto-generated

Comment (BinaryFissionGames) [reply receiver/hostmetricsreceiver/README.md]: Fixed, totally missed my editor formatting this on save 🤦

**Pull request merged**